File: frontend/src/lib/stats.ts
import { apiFetch } from "@/lib/api";

export interface MonthlyIncome {
  month: string;
  amount: number;
}

export interface OccupancyPoint {
  month: string;
  occupied: number;
  total: number;
  rate: number;
}

export interface MaintenanceStatusCount {
  status: string;
  count: number;
}

export interface DashboardStats {
  outstanding: number;
  overdue: number;
  collected_this_month: number;
  properties_total: number;
  properties_occupied: number;
  active_leases: number;
  tenants: number;
  maintenance_open: number;
  monthly_income: MonthlyIncome[];
  occupancy: OccupancyPoint[];
  maintenance_by_status: MaintenanceStatusCount[];
}

export function getDashboardStats() {
  return apiFetch<DashboardStats>("/api/v1/stats");
}

File: backend/app/schemas/stats.py
from decimal import Decimal

from pydantic import BaseModel


class MonthlyIncome(BaseModel):
    month: str
    # A float, not Decimal: Pydantic renders Decimal as a JSON string and the
    # chart cannot plot strings. The money totals above stay Decimal - they are
    # only ever displayed as text.
    amount: float


class OccupancyPoint(BaseModel):
    month: str
    occupied: int
    total: int
    rate: float


class MaintenanceStatusCount(BaseModel):
    status: str
    count: int


class DashboardStats(BaseModel):
    outstanding: Decimal
    overdue: Decimal
    collected_this_month: Decimal
    properties_total: int
    properties_occupied: int
    active_leases: int
    tenants: int
    maintenance_open: int
    monthly_income: list[MonthlyIncome]
    occupancy: list[OccupancyPoint]
    maintenance_by_status: list[MaintenanceStatusCount]
